src/model/immutable/EditorBidiService.js

```diff
@@ -27,25 +27,52 @@ var bidiService;
 var EditorBidiService = {
   getDirectionMap: function(
     content: ContentState,
-    prevBidiMap: ?OrderedMap
+    prevBidiMap: ?OrderedMap,
+    prevContent: ?ContentState,
   ): OrderedMap {
+    if (content === prevContent) {
+      return prevBidiMap;
+    }
+
     if (!bidiService) {
       bidiService = new UnicodeBidiService();
     } else {
       bidiService.reset();
     }
 
-    var blockMap = content.getBlockMap();
-    var nextBidi = blockMap
-      .valueSeq()
-      .map(block => nullthrows(bidiService).getDirection(block.getText()));
-    var bidiMap = OrderedMap(blockMap.keySeq().zip(nextBidi));
+    var newBlockMap = content.getBlockMap();
 
-    if (prevBidiMap != null && Immutable.is(prevBidiMap, bidiMap)) {
-      return prevBidiMap;
-    }
+    if (!prevBidiMap || !prevContent) {
+      var nextBidi = newBlockMap
+        .valueSeq()
+        .map(block => nullthrows(bidiService).getDirection(block.getText()));
 
-    return bidiMap;
+      return OrderedMap(newBlockMap.keySeq().zip(nextBidi));
+    } else {
+      // only update direction for blocks that have changed, then continue looking until we find an unchanged direction.
+
+      var prevLineChanged = false;
+      var prevBlockMap = prevContent.getBlockMap();
+
+      var hasChanged = false;
+
+      var newBidiMap = newBlockMap
+        .map((block, key) => {
+          if (prevLineChanged || block != prevBlockMap.get(key)) {
```

Comment:
I think there are some scenarios where this could fail. One possibility:

```
hello

                  שלום
```

At this point, the second block (which is empty) will be LTR, since it is neutral and depends on the block above. If that block were moved *entirely* below the second block but remained unchanged, it should become RTL by inheriting the direction of the block above it. All three blocks remain otherwise the same.

The conditional here would then never pass, so we would always preserve the direction value from `prevBidiMap`, and the empty block would incorrectly remain LTR.

How expensive is it to map `UnicodeBidiService` over the entire `blockMap`? If it's pretty cheap, then I wonder if we just always do that, keep its results in a regular object, iterate over the existing `bidiMap` and look into the bidi results to find differences, and selectively apply them with set/merge. If any blocks have changed order or been inserted, then we could just do the same `zip` as before. In most cases, it would be an improvement.